# Learning NumPy
NumPy (Numerical Python) is a library for:

- Working with arrays (like lists but faster and more powerful)
- Performing mathematical and statistical operations efficiently
- Handling multi-dimensional data


## Basics

In [ ]:
# installing numpy
!pip install numpy

In [2]:
# importing numpy
import numpy as np

In [3]:
# creating a numpy array
arr = np.array([1, 2, 3]) # 1D array
print(arr)

[1 2 3]


In [4]:
# array attributes
print(arr.shape) #shape of arr, tells the no. of rows and columns
print(arr.ndim) # number of dimensions 1d, 2d etc
print(arr.size) # total no. of elements
print(arr.dtype) # data type of elements

(3,)
1
3
int32


In [5]:
# updating elements
arr[0] = 55 # value at index 0 updating to 55
print(arr)

[55  2  3]


## Special Arrays

In [6]:
# array of zeros
zeros = np.zeros((2, 3)) # 2 rows, 3 cols 
# creates an arr filled with 0.0
# shape is defined as rows, cols
print(zeros)

[[0. 0. 0.]
 [0. 0. 0.]]


In [7]:
# array of ones
ones = np.ones((5, 2)) # rows, cols
print(ones)

[[1. 1.]
 [1. 1.]
 [1. 1.]
 [1. 1.]
 [1. 1.]]


In [8]:
# array with random values
arr_rand = np.random.rand(2, 4) # rows, cols
print(arr_rand)

[[0.45588588 0.79572328 0.97619123 0.34869146]
 [0.62218503 0.02467542 0.91460388 0.94815107]]


## Creating arrays with ranges

In [9]:
# arange()
arr = np.arange(1, 10, 2) # start at 1, end before 10, step by 2
print(arr)

[1 3 5 7 9]


In [10]:
# linspace()
arr = np.linspace(0, 1, 5) # 5 numbers between 0 and 1
print(arr)

# linspace() creates an array with a specific number of evenly spaced points

[0.   0.25 0.5  0.75 1.  ]


## Indexing and Slicing

In [11]:
l = [10, 20, 30, 40, 50]
arr = np.array(l)
print(arr)

[10 20 30 40 50]


In [12]:
# indexing
print(arr[0])
print(arr[-1]) # last element
print(arr[3])

10
50
40


Indexing and slicing works same as python lists

In [13]:
print(arr[1:4])  # Elements from index 1 to 3
print(arr[:3])   # First 3 elements
print(arr[::2])  # Every 2nd element

[20 30 40]
[10 20 30]
[10 30 50]


## Mathematical Operations

In [14]:
# l = [10, 20, 30, 40, 50] already defined earlier
arr = np.array(l)

print(arr + 5) # adds 5 to each element
print(arr * 2) # multiplies each element by 2
print(arr ** 2) # squares each element

[15 25 35 45 55]
[ 20  40  60  80 100]
[ 100  400  900 1600 2500]


In [15]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])

print(arr1 + arr2)
print(arr1 * arr2)

[5 7 9]
[ 4 10 18]


## 2D Arrays

1D array functions tried out in 2D

In [16]:
# creating a 2d array
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8 , 9]])
print(arr)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [17]:
print(arr.shape)
print(arr.ndim)
print(arr.size)
print(arr.dtype)

(3, 3)
2
9
int32


In [18]:
# accessing the elements
print(arr[0, 1]) # element at row 0, col 1
print(arr[0]) # entire row 1
print(arr[: , 2]) # col 2

2
[1 2 3]
[3 6 9]


In [19]:
# slicing
sub_arr = arr[0:2, 1:3] # row 0,1 col 1,2
print(sub_arr)

[[2 3]
 [5 6]]


In [20]:
# updating 
arr[1,2] = 99 # row 1 col 2 element changing
print(arr)

arr[0] = [0, 0, 0] # updating entire row
print(arr)

arr[ : , 2] = [1, 1, 1] # updating col
print(arr)

[[ 1  2  3]
 [ 4  5 99]
 [ 7  8  9]]
[[ 0  0  0]
 [ 4  5 99]
 [ 7  8  9]]
[[0 0 1]
 [4 5 1]
 [7 8 1]]


In [21]:
print(arr + 5) # adds 5 to each element
print(arr * 2) # multiplies each element by 2
print(arr ** 2) # squares each element

[[ 5  5  6]
 [ 9 10  6]
 [12 13  6]]
[[ 0  0  2]
 [ 8 10  2]
 [14 16  2]]
[[ 0  0  1]
 [16 25  1]
 [49 64  1]]


In [22]:
arr1 = np.array([[1, 2, 3], [4, 5, 6], [7, 8 , 9]])
arr2 = np.array([[1, 0, 0], [0, 1, 0], [0, 0 , 1]])

print(arr1 + arr2)
print(arr1 * arr2) # element wise multiplication
print(arr1 @ arr2) # matrix multiplication

[[ 2  2  3]
 [ 4  6  6]
 [ 7  8 10]]
[[1 0 0]
 [0 5 0]
 [0 0 9]]
[[1 2 3]
 [4 5 6]
 [7 8 9]]


Keep an eye on the different types of multiplication

In [23]:
print(arr1.dot(arr2)) # matrix multiplication using dot function
print(np.dot(arr1, arr2)) # matrix multiplication using dot function

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1 2 3]
 [4 5 6]
 [7 8 9]]


## Broadcasting Concept
In simple terms, broadcasting in NumPy is a way of matching array shapes automatically so you can perform operations (like addition, multiplication, etc.) without writing extra code to adjust their sizes manually

For broadcasting to work, the dimensions of the arrays must satisfy these rules:

- The arrays are compared from right to left
- Dimensions are compatible if:
    - They are the same, or
    - One of them is 1

In [24]:
# broadcasting in addition

arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([10, 20, 30])

print(arr1)
print(arr2)
print(arr1 + arr2)

[[1 2 3]
 [4 5 6]]
[10 20 30]
[[11 22 33]
 [14 25 36]]


In [25]:
# broadcasting in addition

arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[10], [20]])

print(arr1)
print(arr2)
print(arr1 + arr2)

[[1 2 3]
 [4 5 6]]
[[10]
 [20]]
[[11 12 13]
 [24 25 26]]


In [26]:
# broadcasting in multiplication

arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([10, 20, 30])

print(arr1 * arr2) # element wise multiplication

[[ 10  40  90]
 [ 40 100 180]]


In [27]:
# broadcasting in multiplication

arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # 2x3
arr2 = np.array([10, 20, 30]) #1x3

print(arr1 @ arr2) # matrix wise multiplication, arr2 becomes 3x1 when multiplying

[140 320]


In [28]:
# broadcasting in multiplication

arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # 2x3
arr2 = np.array([[1, 0, 0], [0, 1, 0], [0, 0 , 1]]) #3x3

print(arr1 @ arr2) # matrix wise multiplication

[[1 2 3]
 [4 5 6]]


| Operation            | Example       | Shape Compatibility   | Result Shape  |
|----------------------|---------------|-----------------------|---------------|
| **Element-Wise**     | `arr1 * arr2` | Same shape or broadcastable | Same as input |
| **Matrix Multiplication** | `arr1 @ arr2` | `(a, b) @ (b, c)`       | `(a, c)`       |


## Reshaping arrays

In [29]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # 2x3

print(arr1.reshape(3, 2)) # reshapes into 3x2
print(arr1.ravel()) # flattens into 1D array
print(arr1.T) # transpose

[[1 2]
 [3 4]
 [5 6]]
[1 2 3 4 5 6]
[[1 4]
 [2 5]
 [3 6]]


## Aggregate Functions

In [30]:
arr = np.array([[1, 2, 3], [44, 55, 66]])

print(np.sum(arr)) # sum of all elements
print(np.mean(arr)) # average of all ele
print(np.max(arr), ",", np.min(arr)) # max and min of all ele
print(np.std(arr)) # standard deviation
print(np.var(arr)) # variance
print(np.median(arr)) # median value
print(np.product(arr)) # product of all ele

171
28.5
66 , 1
27.25649769626807
742.9166666666666
23.5
958320


Check use of axis parameter

In [31]:
# using the axix parameter

print(np.sum(arr, axis=1)) # sum of each row
print(np.sum(arr, axis=0)) # sum of each col

print(np.max(arr, axis=1)) # max of each row
print(np.min(arr, axis=0)) # min of each col

[  6 165]
[45 57 69]
[ 3 66]
[1 2 3]


## Handling Missing Values

In [32]:
# missing values

arr_with_nan = np.array([[1, np.nan, 3], [4, 5, np.nan]])
print(arr_with_nan)

[[ 1. nan  3.]
 [ 4.  5. nan]]


In [33]:
# checking for nan values
print(np.isnan(arr_with_nan))

[[False  True False]
 [False False  True]]


In [34]:
# mean ignoring nan values
print(np.nanmean(arr_with_nan))

3.25


In [35]:
# sum ignoring nan values
print(np.nansum(arr_with_nan))

13.0


# At a glance

\usepackage[a4paper,landscape]{geometry}

| **Name**              | **Description**                                                   | **Input Parameters**                            | **Syntax**                     | **Output**                    | **Return Value**        | **Example**                                                                                      |
|-----------------------|-------------------------------------------------------------------|------------------------------------------------|--------------------------------|--------------------------------|-------------------------|--------------------------------------------------------------------------------------------------|
| **`np.sum`**          | Computes the sum of array elements.                              | `axis` (optional: sum along rows/columns)      | `np.sum(arr, axis)`            | Scalar or array            | Sum of elements         | `np.sum([[1, 2], [3, 4]])` → `10`                                                               |
| **`np.mean`**         | Computes the mean (average) of elements.                        | `axis` (optional: mean along rows/columns)     | `np.mean(arr, axis)`           | Scalar or array            | Average value           | `np.mean([[1, 2], [3, 4]])` → `2.5`                                                             |
| **`np.min` / `np.max`** | Finds the minimum/maximum value in the array.                   | `axis` (optional: min/max along rows/columns)  | `np.min(arr, axis)`            | Scalar or array            | Minimum/maximum value   | `np.min([[1, 2], [3, 4]])` → `1`, `np.max([[1, 2], [3, 4]])` → `4`                              |
| **`np.std`**          | Computes the standard deviation of array elements.              | `axis` (optional)                              | `np.std(arr, axis)`            | Scalar or array            | Standard deviation      | `np.std([[1, 2], [3, 4]])` → `1.118033988749895`                                                |
| **`np.var`**          | Computes the variance of array elements.                        | `axis` (optional)                              | `np.var(arr, axis)`            | Scalar or array            | Variance                | `np.var([[1, 2], [3, 4]])` → `1.25`                                                             |
| **`np.median`**       | Finds the median value of the array.                            | `axis` (optional)                              | `np.median(arr, axis)`         | Scalar or array            | Median value            | `np.median([[1, 2], [3, 4]])` → `2.5`                                                           |
| **`np.prod`**         | Computes the product of array elements.                         | `axis` (optional)                              | `np.prod(arr, axis)`           | Scalar or array            | Product of elements     | `np.prod([[1, 2], [3, 4]])` → `24`                                                              |
| **`np.nanmean`**      | Computes the mean while ignoring NaN values.                   | `axis` (optional)                              | `np.nanmean(arr, axis)`        | Scalar or array            | Mean ignoring NaN       | `np.nanmean([[1, np.nan], [3, 4]])` → `2.6666666666666665`                                      |
| **`np.nansum`**       | Computes the sum while ignoring NaN values.                    | `axis` (optional)                              | `np.nansum(arr, axis)`         | Scalar or array            | Sum ignoring NaN        | `np.nansum([[1, np.nan], [3, 4]])` → `8.0`                                                      |
| **`np.isnan`**        | Checks for NaN values in the array.                             | Array                                          | `np.isnan(arr)`                | Boolean array              | `True` for NaN values   | `np.isnan([1, np.nan, 3])` → `[False, True, False]`                                             |
| **`reshape`**         | Changes the shape of an array.                                  | `new_shape` (tuple)                           | `arr.reshape(new_shape)`       | Reshaped array            | New shape               | `np.array([1, 2, 3, 4]).reshape(2, 2)` → `[[1, 2], [3, 4]]`                                     |
| **`ravel`**           | Flattens the array into 1D.                                     | None                                           | `arr.ravel()`                  | 1D array                  | Flattened array         | `np.array([[1, 2], [3, 4]]).ravel()` → `[1, 2, 3, 4]`                                           |
| **`T` (Transpose)**   | Transposes a 2D array (rows become columns).                   | None                                           | `arr.T`                        | Transposed array          | Transposed view         | `np.array([[1, 2], [3, 4]]).T` → `[[1, 3], [2, 4]]`                                             |
| **Broadcasting**      | Automatically expands smaller arrays to match larger ones for element-wise operations. | N/A                                | `arr1 + arr2`, `arr1 * scalar` | Broadcasted array         | Element-wise result     | `[[1, 2], [3, 4]] * [10, 20]` → `[[10, 40], [30, 80]]`                                          |
| **Matrix Multiplication** | Performs dot product of matrices.                           | `(a, b) @ (b, c)`                             | `arr1 @ arr2`                  | New matrix shape `(a, c)` | Matrix product          | `np.array([[1, 2], [3, 4]]) @ np.array([[5, 6], [7, 8]])` → `[[19, 22], [43, 50]]`              |
| **`np.sum` with Axis** | Sums array elements along a specified axis.                   | `axis=0` (columns) / `axis=1` (rows)          | `np.sum(arr, axis=0)`          | Array reduced by axis     | Summed result           | `np.sum([[1, 2], [3, 4]], axis=0)` → `[4, 6]`                                                   |
| **`np.min` with Axis** | Finds the minimum along rows/columns.                         | `axis=0` (columns) / `axis=1` (rows)          | `np.min(arr, axis=1)`          | Array reduced by axis     | Minimums by axis        | `np.min([[1, 2], [3, 4]], axis=1)` → `[1, 3]`                                                   |


| **Concept**               | **Operator / Function** | **Description**                                                   | **Input Parameters**                            | **Syntax**                     | **Output**                    | **Return Value**        | **Example**                                                                                      |
|---------------------------|-------------------------|-------------------------------------------------------------------|------------------------------------------------|--------------------------------|--------------------------------|-------------------------|--------------------------------------------------------------------------------------------------|
| **Creating Arrays**       | `np.array`             | Creates an array from a list or tuple.                           | `object` (list, tuple, etc.), `dtype` (optional) | `np.array([1, 2, 3])`         | N-dimensional array         | Numpy array             | `np.array([1, 2, 3])` → `[1 2 3]`                                                               |
|                           | `np.zeros`             | Creates an array filled with zeros.                              | `shape` (tuple), `dtype` (optional)             | `np.zeros((2, 3))`             | Array of zeros              | Array filled with zeros | `np.zeros((2, 3))` → `[[0. 0. 0.] [0. 0. 0.]]`                                                  |
|                           | `np.ones`              | Creates an array filled with ones.                               | `shape` (tuple), `dtype` (optional)             | `np.ones((2, 2))`              | Array of ones               | Array filled with ones  | `np.ones((2, 2))` → `[[1. 1.] [1. 1.]]`                                                         |
|                           | `np.arange`            | Creates an array with evenly spaced values.                      | `start`, `stop`, `step`                         | `np.arange(1, 10, 2)`          | 1D array                   | Array with steps         | `np.arange(1, 10, 2)` → `[1 3 5 7 9]`                                                           |
|                           | `np.linspace`          | Creates an array with evenly spaced values between two bounds.   | `start`, `stop`, `num` (default 50)             | `np.linspace(1, 10, 5)`        | 1D array                   | Array with `num` values | `np.linspace(1, 10, 5)` → `[ 1. 3.25 5.5 7.75 10. ]`                                            |
| **Reshaping Arrays**       | `reshape`             | Changes the shape of an array without changing data.             | `new_shape` (tuple)                             | `arr.reshape(new_shape)`       | Reshaped array            | New shape               | `np.array([1, 2, 3, 4]).reshape(2, 2)` → `[[1 2] [3 4]]`                                        |
|                           | `ravel`               | Flattens the array into 1D.                                      | None                                           | `arr.ravel()`                  | 1D array                  | Flattened array         | `np.array([[1, 2], [3, 4]]).ravel()` → `[1, 2, 3, 4]`                                           |
|                           | `T` (Transpose)       | Transposes a 2D array.                                           | None                                           | `arr.T`                        | Transposed array          | Transposed view         | `np.array([[1, 2], [3, 4]]).T` → `[[1 3] [2 4]]`                                                |
| **Element-Wise Operations** | `+`, `-`, `*`, `/`   | Arithmetic operations applied element-wise.                      | Two arrays or array & scalar                    | `arr1 + arr2`, `arr1 * scalar` | Array                     | Element-wise result     | `[[1, 2], [3, 4]] + [[5, 6], [7, 8]]` → `[[6, 8], [10, 12]]`                                    |
|                           | Broadcasting           | Automatically expands smaller arrays for element-wise operations.| N/A                                            | `arr1 + arr2`, `arr * scalar`  | Broadcasted array         | Element-wise result     | `[[1, 2], [3, 4]] * [10, 20]` → `[[10, 40], [30, 80]]`                                          |
| **Matrix Operations**      | `@`                  | Performs matrix multiplication (dot product).                   | Two matrices `(a, b) @ (b, c)`                 | `arr1 @ arr2`                  | Matrix result `(a, c)`    | Matrix product          | `[[1, 2], [3, 4]] @ [[5, 6], [7, 8]]` → `[[19, 22] [43, 50]]`                                   |
| **Aggregate Functions**    | `np.sum`             | Computes the sum of array elements.                              | `axis` (optional: sum along rows/columns)      | `np.sum(arr, axis)`            | Scalar or array            | Sum of elements         | `np.sum([[1, 2], [3, 4]])` → `10`                                                               |
|                           | `np.mean`            | Computes the mean (average) of elements.                        | `axis` (optional: mean along rows/columns)     | `np.mean(arr, axis)`           | Scalar or array            | Average value           | `np.mean([[1, 2], [3, 4]])` → `2.5`                                                             |
|                           | `np.min` / `np.max`   | Finds the minimum/maximum value in the array.                   | `axis` (optional: min/max along rows/columns)  | `np.min(arr, axis)`            | Scalar or array            | Minimum/maximum value   | `np.min([[1, 2], [3, 4]])` → `1`, `np.max([[1, 2], [3, 4]])` → `4`                              |
|                           | `np.std`             | Computes the standard deviation of array elements.              | `axis` (optional)                              | `np.std(arr, axis)`            | Scalar or array            | Standard deviation      | `np.std([[1, 2], [3, 4]])` → `1.118`                                                            |
|                           | `np.median`          | Finds the median value of the array.                            | `axis` (optional)                              | `np.median(arr, axis)`         | Scalar or array            | Median value            | `np.median([[1, 2], [3, 4]])` → `2.5`                                                           |
| **Handling Missing Values** | `np.isnan`          | Checks for NaN values in the array.                             | Array                                          | `np.isnan(arr)`                | Boolean array              | `True` for NaN values   | `np.isnan([1, np.nan, 3])` → `[False, True, False]`                                             |
|                           | `np.nanmean`         | Computes the mean while ignoring NaN values.                   | `axis` (optional)                              | `np.nanmean(arr, axis)`        | Scalar or array            | Mean ignoring NaN       | `np.nanmean([[1, np.nan], [3, 4]])` → `2.6667`                                                  |
|                           | `np.nansum`          | Computes the sum while ignoring NaN values.                    | `axis` (optional)                              | `np.nansum(arr, axis)`         | Scalar or array            | Sum ignoring NaN        | `np.nansum([[1, np.nan], [3, 4]])` → `8.0`                                                      |
